File: sessao-6/tutorial-14-seguranca-ia/exemplos/busca_revisado.ts
/**
 * busca_revisado.ts — Endpoint de busca de pedidos (versão endurecida).
 *
 * Corrige três brechas presentes em busca_gerado.ts:
 *   1. ORDER BY interpolado → validação via allow-list antes de qualquer uso.
 *   2. Regex com bypass (/^[\w\s]+$/) → padrão restrito sem underscore nem wildcards.
 *   3. LIKE por concatenação → filter com includes() parametrizado (não concatenado).
 *
 * Qualquer campo de ordenação ou termo fora dos critérios aceitos é rejeitado
 * com Error antes de chegar à lógica de busca.
 *
 * Execute: npx ts-node busca_revisado.ts
 */

// ---------------------------------------------------------------------------
// Constantes de domínio
// ---------------------------------------------------------------------------

const STATUS_VALIDOS = new Set(["pendente", "processando", "concluido", "cancelado"]);

const CAMPOS_ORDENACAO_PERMITIDOS = new Set([
  "dataCriacao",
  "valorTotal",
  "numeroPedido",
]);

// Padrão restrito: apenas letras (incluindo acentuadas), dígitos e espaço,
// com comprimento máximo explícito. Fecha os caracteres que tornariam um LIKE
// perigoso se concatenado (%, _, ;, aspas, parênteses, underscore via \w).
const PADRAO_TERMO_SEGURO = /^[a-zA-ZÀ-ÿ0-9 ]{1,50}$/;

// ---------------------------------------------------------------------------
// Modelos de dados
// ---------------------------------------------------------------------------

interface FiltroBusca {
  status:    string;
  clienteId: string;
  termo:     string;
}

interface Pedido {
  id:           number;
  numeroPedido: string;
  clienteId:    string;
  status:       string;
  valorTotal:   number;
  dataCriacao:  string;
  descricao:    string;
}

function validarFiltros(filtros: FiltroBusca): void {
  if (!STATUS_VALIDOS.has(filtros.status)) {
    throw new Error(`Status inválido: '${filtros.status}'`);
  }
  if (!filtros.clienteId.trim()) {
    throw new Error("clienteId não pode ser vazio");
  }
  if (filtros.termo && !PADRAO_TERMO_SEGURO.test(filtros.termo)) {
    throw new Error(
      `Termo de busca inválido: '${filtros.termo}'. ` +
      "Use apenas letras e espaços (máx. 50 caracteres).",
    );
  }
}

// ---------------------------------------------------------------------------
// Banco de dados simulado em memória
// ---------------------------------------------------------------------------

const BANCO_PEDIDOS: Pedido[] = [
  { id: 1, numeroPedido: "PED-2026-0001", clienteId: "CLI-100", status: "concluido",   valorTotal:  450.00, dataCriacao: "2026-05-10", descricao: "webcam full hd 1080p" },
  { id: 2, numeroPedido: "PED-2026-0002", clienteId: "CLI-100", status: "concluido",   valorTotal:  980.50, dataCriacao: "2026-05-22", descricao: "adaptador usb thunderbolt" },
  { id: 3, numeroPedido: "PED-2026-0003", clienteId: "CLI-100", status: "processando", valorTotal:  275.00, dataCriacao: "2026-06-01", descricao: "teclado mecânico compacto" },
  { id: 4, numeroPedido: "PED-2026-0004", clienteId: "CLI-200", status: "concluido",   valorTotal: 1200.00, dataCriacao: "2026-05-15", descricao: "notebook premium 15 polegadas" },
  { id: 5, numeroPedido: "PED-2026-0005", clienteId: "CLI-200", status: "pendente",    valorTotal:  330.75, dataCriacao: "2026-06-04", descricao: "mouse ergonômico sem fio" },
];

// ---------------------------------------------------------------------------
// Validação de segurança
// ---------------------------------------------------------------------------

/**
 * Valida que o campo de ordenação pertence ao allow-list.
 *
 * Em query builders e ORMs, ORDER BY frequentemente não aceita parâmetro
 * posicional — a coluna precisa ser interpolada. Esta função garante que
 * apenas campos conhecidos sejam usados na ordenação.
 * Qualquer outro valor é rejeitado antes de chegar à lógica de busca.
 */
function campoOrdenacaoSeguro(ordenacao: string): keyof Pedido {
  const campo = ordenacao.trim();
  if (!CAMPOS_ORDENACAO_PERMITIDOS.has(campo)) {
    throw new Error(
      `Campo de ordenação inválido: '${ordenacao}'. ` +
      `Permitidos: ${[...CAMPOS_ORDENACAO_PERMITIDOS].sort().join(", ")}`,
    );
  }
  return campo as keyof Pedido;
}

/**
 * Executa a busca simulada com filtros validados e ordenação via allow-list.
 * O termo vai como valor ao filter/includes — nunca concatenado na query.
 */
function executarQuery(filtros: FiltroBusca, campo: keyof Pedido): Pedido[] {
  const termo = filtros.termo.toLowerCase();
  const filtrados = BANCO_PEDIDOS.filter(
    (p) =>
      p.status === filtros.status &&
      p.clienteId === filtros.clienteId &&
      (termo === "" || p.descricao.toLowerCase().includes(termo)),
  );
  return [...filtrados].sort((a, b) => {
    const va = a[campo];
    const vb = b[campo];
    if (typeof va === "number" && typeof vb === "number") return va - vb;
    return String(va ?? "").localeCompare(String(vb ?? ""));
  });
}

// ---------------------------------------------------------------------------
// Lógica de busca
// ---------------------------------------------------------------------------

/**
 * Busca pedidos aplicando os filtros fornecidos e ordenando pelo campo indicado.
 *
 * WHERE parametrizado: status e clienteId validados antes do filtro.
 * ORDER BY: campo validado por allow-list antes de qualquer uso na ordenação.
 * Termo: validado por regex restrita; passado como valor (não concatenado).
 * Levanta Error para qualquer entrada fora dos critérios aceitos.
 */
function buscarPedidos(filtros: FiltroBusca, ordenacao: string = "dataCriacao"): Pedido[] {
  validarFiltros(filtros);
  const campo = campoOrdenacaoSeguro(ordenacao);
  return executarQuery(filtros, campo);
}

// ---------------------------------------------------------------------------
// Demo
// ---------------------------------------------------------------------------

function demonstrarBuscaNormal(): void {
  const filtros: FiltroBusca = { status: "concluido", clienteId: "CLI-100", termo: "" };
  const resultados = buscarPedidos(filtros, "valorTotal");

  console.log("Busca normal (status=concluido, cliente=CLI-100, ordem=valorTotal):");
  console.log("  " + "Número".padEnd(16) + "Valor".padStart(10) + "  " + "Data");
  console.log("  " + "-".repeat(44));
  for (const p of resultados) {
    console.log("  " + p.numeroPedido.padEnd(16) + ("R$" + p.valorTotal.toFixed(2)).padStart(10) + "  " + p.dataCriacao);
  }
  console.log();
}

function demonstrarRejeicaoOrdenacaoMaliciosa(): void {
  const filtros: FiltroBusca = { status: "concluido", clienteId: "CLI-100", termo: "" };
  const ordenacaoMaliciosa = "valorTotal DESC, (SELECT 1)";

  console.log("Tentativa com ordenacao maliciosa:");
  console.log("  ordenacao recebida: " + JSON.stringify(ordenacaoMaliciosa));
  try {
    buscarPedidos(filtros, ordenacaoMaliciosa);
    console.log("  FALHOU: a busca executou sem erro — allow-list não funcionou.");
  } catch (erro) {
    console.log("  Bloqueado antes da query: " + (erro as Error).message);
  }
  console.log();
}

function demonstrarRejeicaoCaracteresEspeciais(): void {
  const casos: Array<[string, string]> = [
    ["notebook",       "termo legítimo"],
    ["monitor 34",     "termo com espaço — legítimo"],
    ["x'; DROP TABLE", "payload com aspas e ponto-vírgula"],
    ["%_wildcard",     "wildcards LIKE — rejeitados"],
    ["a".repeat(51),   "termo muito longo — rejeitado"],
  ];

  console.log("Tentativas com termos suspeitos (validação de entrada):");
  console.log("  " + "Termo".padEnd(20) + "Resultado");
  console.log("  " + "-".repeat(56));
  for (const [termo, descricao] of casos) {
    try {
      const filtros: FiltroBusca = { status: "concluido", clienteId: "CLI-100", termo };
      validarFiltros(filtros);
      console.log("  " + JSON.stringify(termo.slice(0, 16)).padEnd(20) + "Aceito (" + descricao + ")");
    } catch {
      console.log("  " + JSON.stringify(termo.slice(0, 16)).padEnd(20) + "Rejeitado (" + descricao + ")");
    }
  }
  console.log();
  console.log("  Mesmo que um termo passe a regex, ele vai como valor ao filter/includes");
  console.log("  — nunca concatenado na string da query.");
  console.log();
}

// ─── Execução ────────────────────────────────────────────────────────────────

console.log("=== Busca de Pedidos — versão endurecida ===\n");

demonstrarBuscaNormal();
demonstrarRejeicaoOrdenacaoMaliciosa();
demonstrarRejeicaoCaracteresEspeciais();

File: sessao-6/tutorial-14-seguranca-ia/exemplos/busca_revisado.py
"""
busca_revisado.py — Endpoint de busca de pedidos (versão endurecida).

Corrige três brechas presentes em busca_gerado.py:
  1. ORDER BY interpolado → validação via allow-list antes de qualquer interpolação.
  2. Regex com bypass (r'^[\\w\\s]+$') → padrão restrito que rejeita palavras-chave SQL.
  3. LIKE por concatenação → parâmetro posicional com o % embutido no valor.

Qualquer input fora dos critérios aceitos é rejeitado com ValueError antes
de tocar na query.

Execute: python3 busca_revisado.py
"""

from __future__ import annotations

import re
import sqlite3
from dataclasses import dataclass
from typing import Any, List


# ---------------------------------------------------------------------------
# Constantes de domínio
# ---------------------------------------------------------------------------

STATUS_VALIDOS = frozenset({"pendente", "processando", "concluido", "cancelado"})

COLUNAS_ORDENACAO_PERMITIDAS = frozenset({
    "data_criacao",
    "valor_total",
    "numero_pedido",
})

# Padrão restrito: apenas letras (incluindo acentuadas), dígitos e espaço,
# com comprimento máximo explícito. O padrão fecha os caracteres especiais que
# tornariam o LIKE perigoso se fosse concatenado (%, _, ;, aspas, parênteses).
# Nota: mesmo que um termo passe aqui, ele vai como parâmetro posicional no LIKE
# — nunca concatenado na query. A regex é defesa em profundidade, não a única barreira.
_PADRAO_TERMO_SEGURO = re.compile(r"^[a-zA-ZáàâãéêíóôõúüçÁÀÂÃÉÊÍÓÔÕÚÜÇ0-9 ]{1,50}$")


# ---------------------------------------------------------------------------
# Modelos de dados
# ---------------------------------------------------------------------------

@dataclass
class FiltroBusca:
    """Filtros para a busca de pedidos."""

    status: str
    cliente_id: str
    termo: str = ""

    def __post_init__(self) -> None:
        if self.status not in STATUS_VALIDOS:
            raise ValueError(f"Status inválido: '{self.status}'")
        if not self.cliente_id.strip():
            raise ValueError("cliente_id não pode ser vazio")
        if self.termo and not _PADRAO_TERMO_SEGURO.match(self.termo):
            raise ValueError(
                f"Termo de busca inválido: '{self.termo}'. "
                "Use apenas letras e espaços (máx. 50 caracteres)."
            )


@dataclass
class Pedido:
    id: int
    numero_pedido: str
    cliente_id: str
    status: str
    valor_total: float
    data_criacao: str
    descricao: str


# ---------------------------------------------------------------------------
# Banco de dados simulado em memória
# ---------------------------------------------------------------------------

def _criar_banco_simulado() -> sqlite3.Connection:
    """Cria e popula um banco SQLite em memória para demonstração."""
    conn = sqlite3.connect(":memory:")
    conn.row_factory = sqlite3.Row
    conn.execute("""
        CREATE TABLE pedidos (
            id            INTEGER PRIMARY KEY,
            numero_pedido TEXT    NOT NULL,
            cliente_id    TEXT    NOT NULL,
            status        TEXT    NOT NULL,
            valor_total   REAL    NOT NULL,
            data_criacao  TEXT    NOT NULL,
            descricao     TEXT    NOT NULL
        )
    """)
    conn.executemany(
        "INSERT INTO pedidos VALUES (?, ?, ?, ?, ?, ?, ?)",
        [
            (1, "PED-2026-0001", "CLI-100", "concluido",   450.00, "2026-05-10", "entrega expressa notebook"),
            (2, "PED-2026-0002", "CLI-100", "concluido",   980.50, "2026-05-22", "monitor ultrawide 34 polegadas"),
            (3, "PED-2026-0003", "CLI-100", "processando", 275.00, "2026-06-01", "teclado mecânico compacto"),
            (4, "PED-2026-0004", "CLI-200", "concluido",  1200.00, "2026-05-15", "notebook premium 15 polegadas"),
            (5, "PED-2026-0005", "CLI-200", "pendente",    330.75, "2026-06-04", "mouse ergonômico sem fio"),
        ],
    )
    conn.commit()
    return conn


_conn = _criar_banco_simulado()


# ---------------------------------------------------------------------------
# Validação de segurança
# ---------------------------------------------------------------------------

def _coluna_ordenacao_segura(ordenacao: str) -> str:
    """
    Valida que a coluna de ordenação pertence ao allow-list.

    ORDER BY não pode ser parametrizado com ? na maioria dos drivers SQL.
    Esta função garante que apenas identificadores conhecidos sejam
    interpolados na query — qualquer outro valor é rejeitado antes de
    chegar ao banco.
    """
    coluna = ordenacao.strip().lower()
    if coluna not in COLUNAS_ORDENACAO_PERMITIDAS:
        raise ValueError(
            f"Coluna de ordenação inválida: '{ordenacao}'. "
            f"Permitidas: {sorted(COLUNAS_ORDENACAO_PERMITIDAS)}"
        )
    return coluna


# ---------------------------------------------------------------------------
# Lógica de busca
# ---------------------------------------------------------------------------

def buscar_pedidos(
    filtros: FiltroBusca,
    ordenacao: str = "data_criacao",
) -> List[dict[str, Any]]:
    """
    Busca pedidos aplicando os filtros fornecidos e ordenando pelo campo indicado.

    WHERE parametrizado com ? para status, cliente_id e o padrão LIKE.
    O % do LIKE é embutido no valor do parâmetro — não concatenado na query.
    ORDER BY validado por allow-list antes de qualquer interpolação.
    Levanta ValueError para qualquer entrada fora dos critérios aceitos.
    """
    coluna = _coluna_ordenacao_segura(ordenacao)
    termo_like = f"%{filtros.termo}%" if filtros.termo else "%"
    query = (
        "SELECT id, numero_pedido, cliente_id, status, valor_total, data_criacao, descricao "
        "FROM pedidos "
        "WHERE status = ? AND cliente_id = ? AND descricao LIKE ? "
        f"ORDER BY {coluna}"
    )
    cursor = _conn.execute(query, (filtros.status, filtros.cliente_id, termo_like))
    return [dict(row) for row in cursor.fetchall()]


# ---------------------------------------------------------------------------
# Demo
# ---------------------------------------------------------------------------

def _demonstrar_busca_normal() -> None:
    filtros = FiltroBusca(status="concluido", cliente_id="CLI-100", termo="")
    resultados = buscar_pedidos(filtros, ordenacao="valor_total")

    print("Busca normal (status=concluido, cliente=CLI-100, ordem=valor_total):")
    print(f"  {'Número':<16} {'Valor':>10}  {'Data'}")
    print("  " + "-" * 44)
    for p in resultados:
        print(f"  {p['numero_pedido']:<16} R${p['valor_total']:>9.2f}  {p['data_criacao']}")
    print()


def _demonstrar_rejeicao_ordenacao_maliciosa() -> None:
    filtros = FiltroBusca(status="concluido", cliente_id="CLI-100")
    ordenacao_maliciosa = "valor_total DESC, (SELECT 1)"

    print("Tentativa com ordenacao maliciosa:")
    print(f"  ordenacao recebida: {ordenacao_maliciosa!r}")
    try:
        buscar_pedidos(filtros, ordenacao=ordenacao_maliciosa)
        print("  FALHOU: a query executou sem erro — allow-list não funcionou.")
    except ValueError as erro:
        print(f"  Bloqueado antes da query: {erro}")
    print()


def _demonstrar_rejeicao_caracteres_especiais() -> None:
    """
    Demonstra que a regex restrita bloqueia os caracteres que tornariam
    o LIKE perigoso — e que, mesmo para termos aceitos, o valor vai como
    parâmetro posicional (nunca concatenado na query).
    """
    print("Tentativas com termos suspeitos (validação de entrada):")
    print(f"  {'Termo':<20} Resultado")
    print("  " + "-" * 56)
    casos = [
        ("notebook",       "termo legítimo"),
        ("monitor 34",     "termo com espaço — legítimo"),
        ("x'; DROP TABLE", "payload com aspas e ponto-vírgula"),
        ("%_wildcard",     "wildcards LIKE — rejeitados"),
        ("a" * 51,         "termo muito longo — rejeitado"),
    ]
    for termo, descricao in casos:
        try:
            FiltroBusca(status="concluido", cliente_id="CLI-100", termo=termo)
            print(f"  {repr(termo[:18]):<20} Aceito ({descricao})")
        except ValueError:
            print(f"  {repr(termo[:18]):<20} Rejeitado ({descricao})")
    print()
    print("  Mesmo que um termo passe a regex, ele vai como parâmetro '?'")
    print("  no LIKE — nunca concatenado na string da query.")
    print()


def _demonstrar_rejeicao_coluna_inexistente() -> None:
    filtros = FiltroBusca(status="concluido", cliente_id="CLI-100")

    print("Tentativa com coluna inexistente:")
    print("  ordenacao recebida: 'cpf_cliente'")
    try:
        buscar_pedidos(filtros, ordenacao="cpf_cliente")
        print("  FALHOU: a query executou sem erro — allow-list não funcionou.")
    except ValueError as erro:
        print(f"  Bloqueado antes da query: {erro}")
    print()


if __name__ == "__main__":
    print("=== Busca de Pedidos — versão endurecida ===\n")

    _demonstrar_busca_normal()
    _demonstrar_rejeicao_ordenacao_maliciosa()
    _demonstrar_rejeicao_caracteres_especiais()
    _demonstrar_rejeicao_coluna_inexistente()
